42/README.md

```diff
@@ -70,14 +71,129 @@ The above process ensure the node keeps validating and accepting new blocks whil
+3. Replay the whole blockchain after a migration.
+4. Join the network as a new validator after a migration.
+
+Additionally, BigchainDB HTTP API usage has to be seamless - all the HTTP responses has to look like there were no migrations.
+
+Below we describe how to perform a migration while meeting the postulated requirements. Note that there is no limitation on the number of consequent migrations that can be performed this way.
+
+#### 1. Stop building blocks
+
+A CLI command can be used to instruct the node to stop building blocks at a chosen height.
+
+To make sure no blocks are committed after the agreed height, we propose an election process:
+
+- The initiator creates an election, which includes a block height.
+- The validators vote for the election. Once the election is concluded, new blocks are not committed by the validators. New transactions from the users are rejected. New blocks sent by Tendermint are rejected too. The conditions for resuming the chain operation are described below.
+
+To perform a migration election, we propose 2 CLI commands.
+
+The initiator executes:
+
+```
+$ bigchaindb migration new <height> --private-key /home/user/.tendermint/config/priv_validator.json
+```
+
+The command outputs the migration ID. The initiator distributes it among other members of the network. The process is similar to adding new validators.
+
+Validators vote for the migration:
+
+```
+$ bigchaindb migration approve <migration-id> --private-key /home/user/.tendermint/config/priv_validator.json
+```
+
+Validators can watch how the election goes:
+
+```
+$ bigchaindb migration status <migration-id>
+```
+
+Outputs:
+```
+votes_recieved=<Sum_of_votes_recieved>
+votes_allocated=<Sum_of_votes_allocated_in_election>
+network_size=<Total_network_power>
+```
+
+#### 2. Replay the blockchain
+
+In order to replay the chain up to the migration height, we need to either keep the old Tendermint version running or skip replaying the corresponding part of the chain = archive the chain. We consider the former to be a huge hassle so we conceive an archiving scheme further in this chapter.
+
+#### 3. Start the new chain
+
+Validators have to install and launch the new version of Tendermint. They need to prepare a new `genesis.json`. The new genesis file has to contain the validator set, the application hash (both at the migration height), and the identifier of the new Tendermint chain (`chain_id`).
+
+The application hash is calculated by computing the Merkle tree root of the current UTXO.
+
+`chain_id` is generated and stored by BigchainDB upon conclusion of the migration election. When Tendermint sends this ID as part of the `InitChain` ABCI request, BigchainDB understands that the user has switched to a new Tendermint version so BigchainDB switches to accept new transactions and blocks.
+
+To offer a convenient way to get the data, we propose a CLI command:
+
+```
+$ bigchaindb status
```

Comment:
The command seems very general i.e. just by looking at the command is not clear that it would return the migration data. I would prefer a different name i.e. something like `bighciandb migration get-genesis <migration-id>`?

Reply:
My thinking is `bigchaindb status` can be as well useful outside of the migration scope. For instance, it is a good way for the operator to know the current validator set.

I am also thinking to change `bigchaindb migration dump/apply` to `bigchaindb dump/apply`. This way, node operators can back things up every once in a while so that if their storage experiences a critical failure, they can replay the chain from the dump. In the future, we may include Tendermint's `genesis.json` there (for instance, after we package a single binary).

Does it sound good?

Reply:
I would prefer a separation of concern i.e. one command only does one thing. If we choose `bigchaindb status` to list the genesis of the next migration then it can only do that nothing else.

Reply:
The idea is to make `bigchaindb status` show the current state of the system, similarly to what the Tendermint's RPC endpoint `/status` is doing.

Figuring out the genesis of the migrated chain is still a separate concern which would in this case involve using the generic status command.

Basically, my point is it might be inconvenient to create a dedicated command for every small concern. Imagine the amount of commands we might have - `bigchaindb get-validators`, `bigchaindb get-chain-id`, `bigchaindb migration get-validators`, `bigchaindb some-integration get-validators`, etc.

On a side note, it's not a good experience to copy and paste the data. Hopefully, in the near future we will have control over the Tendermint setup so that users can simply issue a `bigchaindb continue` type of command after downloading the new binary - it will then automatically pick up genesis data.

What do you think about `bigchaindb dump/apply` as a mean to replay the chain from a "checkpoint", useful even if there were no migrations?

Reply:
Pondered about dumping a little more. I would keep `bigchaindb migration dump` as a command for dumping the state at the last migration height. A generic `bigchaindb dump` might still be introduced in the future but it will take a snapshot as of the last known height or at the given height, when specified.

Loading the dump is height-agnostic though so I would prefer a generic `bigchaindb apply <dump>` command.

Alternative names may be `bigchaindb migration create-snapshot`, `bigchaindb create-snapshot` (to be introduced later, not now), and `bigchaindb load-from-snapshot`.

What do you think?

Reply:
Discussed offline - migration snapshots are only concerned with MongoDB so we can just use `mongodump`/`mongorestore` at this moment.

Reply:
The BEP should specify the output of `bigchaindb status` command when there is no migration.